### PR TITLE
Feature: support update model metadata API

### PR DESCRIPTION
**Pull request by wwwy3y3**
## Description
* Support update model metadata API: `updateModelMetadata` mutation
* Add description field to relationship in `Diagram` query
Screenshot 2024-05-03 at 9 53 06 AM
 
## Note
* If users give empty string, backend will set the field to null in database

## MDL Change
* Adding `description` to model, column properties
* Adding `properties` to relationship, which will include description.
<details>

<summary>MDL</summary>

```js
{
  "schema": "public",
  "catalog": "wrenai",
  "models": [
    {
      "name": "customers",
      "columns": [
        {
          "name": "City",
          "type": "VARCHAR",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "The Customer City, where the customer company is located. Also called \"customer segment\".",
            "displayName": "City"
          }
        },
        {
          "name": "Id",
          "type": "VARCHAR",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "This is a unique identifier for each customer in the data model.",
            "displayName": "顧客 ID (測試)"
          }
        },
        {
          "name": "State",
          "type": "VARCHAR",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A field indicating the state where the customer is located.",
            "displayName": "State"
          }
        },
        {
          "name": "orders",
          "type": "orders",
          "properties": null,
          "relationship": "CustomersOrders",
          "isCalculated": false,
          "notNull": false
        }
      ],
      "refSql": "select * from main.customers",
      "cached": 0,
      "refreshTime": null,
      "properties": {
        "schema": "main",
        "catalog": "memory",
        "description": "customer information testing testing testing",
        "displayName": "顧客 testing"
      },
      "primaryKey": ""
    },
    {
      "name": "order_items",
      "columns": [
        {
          "name": "FreightValue",
          "type": "DOUBLE",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A numerical value representing the cost of shipping for an item in an order.",
            "displayName": "FreightValue"
          }
        },
        {
          "name": "ItemNumber",
          "type": "BIGINT",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "The sequential number of the order item in this order. Each order item in an order has its unique ItemNumber.",
            "displayName": "ItemNumber"
          }
        },
        {
          "name": "OrderId",
          "type": "VARCHAR",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A VARCHAR value indicating the order that this order_item belongs to. The column is used to map the order_item to Orders model in the OrdersOrder_items relationship.",
            "displayName": "OrderId"
          }
        },
        {
          "name": "Price",
          "type": "DOUBLE",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A numerical value representing the price of an item in an order.",
            "displayName": "Price"
          }
        },
        {
          "name": "ProductId",
          "type": "VARCHAR",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A VARCHAR value representing the product of this order_item. The column is used to map the order_item to Products model using ProductsOrder_items relationship.",
            "displayName": "ProductId"
          }
        },
        {
          "name": "ShippingLimitDate",
          "type": "DATE",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A date value indicating the limit by which an item should be shipped according to the order. It helps track the deadline for shipping items in the \"order_items\" model.",
            "displayName": "ShippingLimitDate"
          }
        },
        {
          "name": "orders",
          "type": "orders",
          "properties": null,
          "relationship": "OrdersOrder_items",
          "isCalculated": false,
          "notNull": false
        },
        {
          "name": "products",
          "type": "products",
          "properties": null,
          "relationship": "ProductsOrder_items",
          "isCalculated": false,
          "notNull": false
        }
      ],
      "refSql": "select * from main.order_items",
      "cached": 0,
      "refreshTime": null,
      "properties": {
        "schema": "main",
        "catalog": "memory",
        "description": "The model is used to store information about items in orders, including details like prices, product IDs, shipping limits, and relationships with orders and products tables.",
        "displayName": "order_items"
      },
      "primaryKey": ""
    },
    {
      "name": "orders",
      "columns": [
        {
          "name": "ApprovedTimestamp",
          "type": "TIMESTAMP",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A column that represents the timestamp when the order was approved.",
            "displayName": "ApprovedTimestamp"
          }
        },
        {
          "name": "CustomerId",
          "type": "VARCHAR",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A unique identifier representing the customer who purchased this order.",
            "displayName": "CustomerId"
          }
        },
        {
          "name": "DeliveredCarrierDate",
          "type": "DATE",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A column that represents the date when the order was delivered by the carrier.",
            "displayName": "DeliveredCarrierDate"
          }
        },
        {
          "name": "DeliveredCustomerDate",
          "type": "DATE",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A column that represents the date when the order was delivered to the customer.",
            "displayName": "DeliveredCustomerDate"
          }
        },
        {
          "name": "EstimatedDeliveryDate",
          "type": "DATE",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A column that represents the estimated delivery date of the order.",
            "displayName": "EstimatedDeliveryDate"
          }
        },
        {
          "name": "OrderId",
          "type": "VARCHAR",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A column that represents a unique identifier of this order.",
            "displayName": "OrderId"
          }
        },
        {
          "name": "PurchaseTimestamp",
          "type": "TIMESTAMP",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A column that represents the timestamp when the order was purchased.",
            "displayName": "PurchaseTimestamp"
          }
        },
        {
          "name": "Status",
          "type": "VARCHAR",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A column representing the status of the order.",
            "displayName": "Status"
          }
        },
        {
          "name": "customers",
          "type": "customers",
          "properties": null,
          "relationship": "CustomersOrders",
          "isCalculated": false,
          "notNull": false
        },
        {
          "name": "order_items",
          "type": "order_items",
          "properties": null,
          "relationship": "OrdersOrder_items",
          "isCalculated": false,
          "notNull": false
        },
        {
          "name": "reviews",
          "type": "reviews",
          "properties": null,
          "relationship": "OrdersReviews",
          "isCalculated": false,
          "notNull": false
        },
        {
          "name": "payments",
          "type": "payments",
          "properties": null,
          "relationship": "PaymentsOrders",
          "isCalculated": false,
          "notNull": false
        }
      ],
      "refSql": "select * from main.orders",
      "cached": 0,
      "refreshTime": null,
      "properties": {
        "schema": "main",
        "catalog": "memory",
        "description": "A model representing the orders data.",
        "displayName": "orders"
      },
      "primaryKey": ""
    },
    {
      "name": "payments",
      "columns": [
        {
          "name": "Installments",
          "type": "BIGINT",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A column representing the number of installments in the payments data model.",
            "displayName": "Installments"
          }
        },
        {
          "name": "OrderId",
          "type": "VARCHAR",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A column representing the order id associated with this payment. The column is used to map the payment to the order using PaymentsOrders relationship.",
            "displayName": "OrderId"
          }
        },
        {
          "name": "Sequential",
          "type": "BIGINT",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A column representing the sequential number of the payment in its corresponding order. Each payment in the order has its unique sequential number.",
            "displayName": "Sequential"
          }
        },
        {
          "name": "Type",
          "type": "VARCHAR",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A column representing the type of payment in the payments data model.",
            "displayName": "Type"
          }
        },
        {
          "name": "Value",
          "type": "DOUBLE",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A column representing the value of the payment in the payments data model.",
            "displayName": "Value"
          }
        },
        {
          "name": "orders",
          "type": "orders",
          "properties": null,
          "relationship": "PaymentsOrders",
          "isCalculated": false,
          "notNull": false
        }
      ],
      "refSql": "select * from main.payments",
      "cached": 0,
      "refreshTime": null,
      "properties": {
        "schema": "main",
        "catalog": "memory",
        "description": "A model representing the payment records, including installments, order IDs, sequential numbers, payment types, values, and relationships with orders.",
        "displayName": "payments"
      },
      "primaryKey": ""
    },
    {
      "name": "products",
      "columns": [
        {
          "name": "Category",
          "type": "VARCHAR",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A category that classifies the products in the data model.",
            "displayName": "Category"
          }
        },
        {
          "name": "Id",
          "type": "VARCHAR",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A unique identifier assigned to each product in the data model.",
            "displayName": "Id"
          }
        },
        {
          "name": "Name",
          "type": "VARCHAR",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A name of the product in the data model.",
            "displayName": "Name"
          }
        },
        {
          "name": "order_items",
          "type": "order_items",
          "properties": null,
          "relationship": "ProductsOrder_items",
          "isCalculated": false,
          "notNull": false
        }
      ],
      "refSql": "select * from main.products",
      "cached": 0,
      "refreshTime": null,
      "properties": {
        "schema": "main",
        "catalog": "memory",
        "description": "A data model containing information about products such as category, ID, and name, with a relationship to order items.",
        "displayName": "products"
      },
      "primaryKey": ""
    },
    {
      "name": "reviews",
      "columns": [
        {
          "name": "AnswerTimestamp",
          "type": "TIMESTAMP",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "The date when the answer was provided.",
            "displayName": "AnswerTimestamp"
          }
        },
        {
          "name": "CreationTimestamp",
          "type": "TIMESTAMP",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "The date when the review was created.",
            "displayName": "CreationTimestamp"
          }
        },
        {
          "name": "Id",
          "type": "VARCHAR",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "A unique identifier assigned to each review entry.",
            "displayName": "Id"
          }
        },
        {
          "name": "OrderId",
          "type": "VARCHAR",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "The order id of the order which the review belongs to.",
            "displayName": "OrderId"
          }
        },
        {
          "name": "Score",
          "type": "BIGINT",
          "isCalculated": 0,
          "notNull": 0,
          "expression": "",
          "properties": {
            "description": "The score associated with each review entry.",
            "displayName": "Score"
          }
        },
        {
          "name": "orders",
          "type": "orders",
          "properties": null,
          "relationship": "OrdersReviews",
          "isCalculated": false,
          "notNull": false
        }
      ],
      "refSql": "select * from main.reviews",
      "cached": 0,
      "refreshTime": null,
      "properties": {
        "schema": "main",
        "catalog": "memory",
        "description": "A model containing information about review of orders.",
        "displayName": "reviews"
      },
      "primaryKey": ""
    }
  ],
  "relationships": [
    {
      "name": "CustomersOrders",
      "models": [
        "customers",
        "orders"
      ],
      "joinType": "ONE_TO_MANY",
      "condition": "\"customers\".Id = \"orders\".CustomerId",
      "properties": {}
    },
    {
      "name": "OrdersOrder_items",
      "models": [
        "orders",
        "order_items"
      ],
      "joinType": "ONE_TO_MANY",
      "condition": "\"orders\".OrderId = \"order_items\".OrderId",
      "properties": {}
    },
    {
      "name": "ProductsOrder_items",
      "models": [
        "products",
        "order_items"
      ],
      "joinType": "ONE_TO_MANY",
      "condition": "\"products\".Id = \"order_items\".ProductId",
      "properties": {}
    },
    {
      "name": "OrdersReviews",
      "models": [
        "orders",
        "reviews"
      ],
      "joinType": "ONE_TO_MANY",
      "condition": "\"orders\".OrderId = \"reviews\".OrderId",
      "properties": {
        "description": "which customer bought this order"
      }
    },
    {
      "name": "PaymentsOrders",
      "models": [
        "payments",
        "orders"
      ],
      "joinType": "ONE_TO_MANY",
      "condition": "\"payments\".OrderId = \"orders\".OrderId",
      "properties": {}
    }
  ],
  "views": []
}
```

</details>

### CODE REVIEW

File: wren-ui/src/apollo/server/resolvers/modelResolver.ts [hunk 1]
```diff
@@ -338,6 +469,21 @@ export class ModelResolver {
...
+    // so we return null
+    if (value === '') {
+      return null;
+    }
+
+    // otherwise, return the value
+    return value;
+  }
+
+  private isMetadataValueEmpty(metadata: Record<string, string>) {
```
Comment (onlyjackfrost) [inline wren-ui/src/apollo/server/resolvers/modelResolver.ts:483]: Seems this method was not used.

File: wren-ui/src/apollo/server/types/diagram.ts [hunk 2]
```diff
@@ -70,4 +70,5 @@ export interface DiagramModelRelationField {
   fromColumnName: string;
   toModelName: string;
   toColumnName: string;
+  description: string;
```
Comment (onlyjackfrost) [inline wren-ui/src/apollo/server/types/diagram.ts:73]: Should the description be returned In the `transformModelRelationField` method?

**Review by onlyjackfrost — APPROVED** [review]
Besides comments, others LGTM

File: wren-ui/src/apollo/server/resolvers/modelResolver.ts (hunk 1 above)
Comment (wwwy3y3) [reply wren-ui/src/apollo/server/resolvers/modelResolver.ts]: addressed in f3e2106

File: wren-ui/src/apollo/server/types/diagram.ts (hunk 2 above)
Comment (wwwy3y3) [reply wren-ui/src/apollo/server/types/diagram.ts:73]: Already returned here I think

https://github.com/Canner/WrenAI/blob/41d38154ee4a0c73f7ade77703e7882dcf2f96cb/wren-ui/src/apollo/server/resolvers/diagramResolver.ts#L149

**Review by onlyjackfrost — APPROVED** [review]
LGTM

**Pull request merged**